js/lightbeam.js

```diff
@@ -1,9 +1,63 @@
+// transforms the object of nested objects 'websites' into a
+// usable format for d3
+function transformData(websites) {
+  const nodes = [];
+  const links = [];
+  const thirdParties = [];
+  // make first party nodes
+  for (const firstParty in websites) {
+    const node = {};
+    node.hostname = firstParty;
+    node.favicon = websites[firstParty].faviconUrl || '';
+    node.party = 'first';
+    nodes.push(node);
+    // make third party nodes and links
+    for (const thirdParty in websites[firstParty].thirdPartyRequests) {
+      // make sure there is only one node per third party
+      if (thirdParties.indexOf(thirdParty) === -1) {
+        // make a third party node
+        thirdParties.push(thirdParty);
+        const node = {};
+        node.hostname = thirdParty;
+        // @todo get third party favicons
+        node.favicon = '';
+        node.party = 'third';
+        nodes.push(node);
+      }
+      const link = {};
+      link.source = firstParty;
+      link.target = thirdParty;
+      links.push(link);
+    }
+  }
+
+  return [nodes, links];
+}
+
 async function renderGraph() {
-  const canvas = document.getElementById('canvas');
-  const context = canvas.getContext('2d');
-  const websites = await storeChild.getAll();
+  let websites = await storeChild.getAll();
+  let data = transformData(websites);
+  let nodesPrev = data[0];
+  let linksPrev = data[1];
 
-  viz.draw(context, websites);
+  viz.init(nodesPrev, linksPrev);
+
+  // when data changes, update
+  // @todo find a good way to listen for changes
+  d3.interval(async function() {
+    websites = await storeChild.getAll();
```

Comment:
I agree about that (also there is the built-in `setInterval` method which does the exact same thing). I have removed this portion from the code, so that currently the page only updates the visualization on page load.

@jonathanKingston , when you say make a bug, are you referring to the storage callback suggestion? Could you elaborate a bit on what that means? I created Issue #75.